src/encoded/static/components/matrix_chip_seq.js

```diff
@@ -750,12 +772,29 @@ class ChIPSeqMatrixPresentation extends React.Component {
 
         const count = getSearchResultCount(chIPSeqData);
         PubSub.publish(SEARCH_RESULT_COUNT_PUBSUB, count);
-        showSpinner(true);
 
-        const updateState = () => this.setState({ chIPSeqData }, () => {
-            showSpinner(false);
+        const isMatrixLarge = isMatrixUpdateLarge(chIPSeqData, this.state.chIPSeqData);
+
+        this.setState({ spinnerActive: true }, () => {
+            if (!isMatrixLarge) {
+                this.setState({ chIPSeqData: null }, () => {
+                    this.setState({ chIPSeqData }, () => {
+                        this.setState({ spinnerActive: false });
+                    });
+                });
+            } else {
+                // defer is used to allow painting of DOM after matrix is drawn. It is hacky.
+                _.defer(() => {
+                    this.setState({ chIPSeqData: null }, () => {
+                        this.setState({ chIPSeqData }, () => {
+                            _.defer(() => {
+                                this.setState({ spinnerActive: false });
+                            });
+                        });
+                    });
+                });
+            }
```

Comment:
Calling setState in the second parameter to setState worries me a little, but I haven’t been able to find anything that says you _can’t_ do this. I’ll assume it’s fine, but good to note unless you found something that says this is fine.

Reply:
That and calling nested-defer's. I have not seen anyone else do these. However, I never read anyone that said don't and the code seems to have no trouble with it. It is not normally good to be that adventurous... But I had to get creative.